# Исследование надёжности заёмщиков

Задача от кредитного отдела банка. Необходимо определить зависимость между семейным положенем, колличеством детей и  фактом погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

## Импорт библиотек и загрузка данных

In [1]:
# Импортируем библиотеку pandas и сохраняем датафрейм
import pandas as pd
df = pd.read_csv('data.csv')
df.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [2]:
#Просматриваем общую информацию о датафрейме
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


- *children —* количество детей в семье
- *days_employed* — общий трудовой стаж в днях
- *dob_years* — возраст клиента в годах
- *education* — уровень образования клиента
- *education_id* — идентификатор уровня образования
- *family_status* — семейное положение
- *family_status_id* — идентификатор семейного положения
- *gender* — пол клиента
- *income_type* — тип занятости
- *debt* — имел ли задолженность по возврату кредитов
- *total_income* — ежемесячный доход
- *purpose* — цель получения кредита

In [3]:
for col in df.columns:
    print(str(col))
    display(df[col].value_counts())

children


 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

days_employed


-327.685916     1
-1580.622577    1
-4122.460569    1
-2828.237691    1
-2636.090517    1
               ..
-7120.517564    1
-2146.884040    1
-881.454684     1
-794.666350     1
-3382.113891    1
Name: days_employed, Length: 19351, dtype: int64

dob_years


35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

education


среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

education_id


1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

family_status


женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

family_status_id


0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

gender


F      14236
M       7288
XNA        1
Name: gender, dtype: int64

income_type


сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64

debt


0    19784
1     1741
Name: debt, dtype: int64

total_income


169846.427535    1
257737.077768    1
200508.675866    1
106196.235958    1
248730.171354    1
                ..
175057.266090    1
101516.604975    1
239154.168013    1
165009.733021    1
189255.286637    1
Name: total_income, Length: 19351, dtype: int64

purpose


свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

Данные из файла прочитаны. Просмотрев общую информацию видим что в датафрейме отсутствует столбцец с id, присутствуют пропуски в столбцах days_employed и total_income.  
Так же мы обнаружили следующие ошибки: 
    - отрицательное значение в стролбце children (возможна ошибка формы или обработки)
    - отрицательное значение в стролбце days_employed (возможна ошибка формы или обработки)
    - значение 0 в стролбце dob_years (ошибка обработки формы или неверное заполнение)
    - разный регистр в столбце education (Свободное заполнение этого пункта в форме/ сохранение исходного заполнения)
    - пол XNA в столбце gender (ошибка заполнения/повреждение данных)

## Предобработка данных

### Обработка пропусков

Из df.info() видим что в датафрейме 21525 строк и 12 столбцов. Обращаем внимание на пропущенные значения в столбцах days_employed и total_income.

In [4]:
df['days_employed'].isnull().sum()

2174

In [5]:
df['total_income'].isnull().sum()

2174

В столбцах days_employed и total_income одинаковое кол-во пропусков. Сравним как эти пропуски соотносятся: методом concat соединим их в одну таблицу и удалим дубликаты. Если они одинаковые то результатом будет пустой датафрейм.

In [6]:
result = pd.concat([df[df['days_employed'].isnull()],df[df['total_income'].isnull()]]).drop_duplicates(keep=False)
print(result)

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []


Пропусков у нас 2174 шт и совпадают по отсутствию данных в столбцах days_employed и total_income. На данное соответствие пропусков следуетобратить внимание, так как присутствует прямая зависимость.

Чтобы понять случайны ли пропуски сопоставим количество анкет по категориальному признаку(income_type) методом value_counts:

In [7]:
df['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64

Доли распределения анкет по должностям.

In [8]:
df['income_type'].value_counts(normalize=True)

сотрудник          0.516562
компаньон          0.236237
пенсионер          0.179141
госслужащий        0.067782
предприниматель    0.000093
безработный        0.000093
студент            0.000046
в декрете          0.000046
Name: income_type, dtype: float64

Сопоставим количество анкет с пропущенными значениями по должности. 

In [9]:
df[df['total_income'].isnull()]['income_type'].value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

Найдем соотношение анкет с проппусками по профессиям в процентах.

In [10]:
df[df['total_income'].isnull()]['income_type'].value_counts(normalize=True)

сотрудник          0.508280
компаньон          0.233671
пенсионер          0.189972
госслужащий        0.067617
предприниматель    0.000460
Name: income_type, dtype: float64

Вывод: так как соотшение анкет по типу занятости вцелом и у анкет с пропусками одинаковое, делаем вывод что пропуски в нашем случае носят случайный характер, тк соотвествуют общему распределению анкет.
Для исследований стаж нам не интересен, поэтому просто заполним пропуски 0. 
Данные о ежемесячном доходе заполним медианными значениями по типу занятости.

Проверим данные по анкетам, где опыт работы больше 0.
Просмотрев таблицу видим что встречаются анкеты где значение опыта работы в днях превышает кол-во полных лет. 
Сравним данные с этим условием с таблицой где опыт работы больше 0.

In [41]:
df[df['dob_years']*365 < df['days_employed']].sample(10)

children  days_employed  dob_years  education  education_id  \
8272          0         334837         49    среднее             1   
2118          0         349741         63    среднее             1   
10115         0         381598         64    среднее             1   
2789          0         376924         61     высшее             0   
19964         1         357913         41    среднее             1   
18598         0         351079         70  начальное             3   
5241          0         337704         61    среднее             1   
15611         0         353634         56    среднее             1   
11569         0         367763         48    среднее             1   
12528         0         330226         59    среднее             1   

          family_status  family_status_id gender income_type  debt  \
8272    женат / замужем                 0      F   пенсионер     0   
2118    женат / замужем                 0      M   пенсионер     0   
10115    вдовец / вдова                 2      F   пенсионер     0   
2789    женат / замужем                 0      F   пенсионер     0   
19964  гражданский брак                 1      F   пенсионер     0   
18598  гражданский брак                 1      F   пенсионер     0   
5241     вдовец / вдова                 2      M   пенсионер     1   
15611   женат / замужем                 0      F   пенсионер     1   
11569  гражданский брак                 1      M   пенсионер     0   
12528    вдовец / вдова                 2      F   пенсионер     0   

       total_income                           purpose  \
8272         149472                     покупка жилья   
2118          45837           покупка жилья для семьи   
10115        127962  сделка с подержанным автомобилем   
2789         282485                высшее образование   
19964        356311  строительство жилой недвижимости   
18598         77684                           свадьба   
5241          63600             получение образования   
15611         59184                     покупка жилья   
11569        121375   операции со своей недвижимостью   
12528        132161                        автомобили   

                                                  lemmas  \
8272                             [покупка,  , жилье, \n]   
2118           [покупка,  , жилье,  , для,  , семья, \n]   
10115  [сделка,  , с,  , подержанный,  , автомобиль, \n]   
2789                       [высокий,  , образование, \n]   
19964     [строительство,  , жилой,  , недвижимость, \n]   
18598                                      [свадьба, \n]   
5241                     [получение,  , образование, \n]   
15611                            [покупка,  , жилье, \n]   
11569    [операция,  , со,  , свой,  , недвижимость, \n]   
12528                                   [автомобиль, \n]   

             purpose_category   income_level  
8272                  ипотека  выше среднего  
2118                  ипотека      маленький  
10115              автокредит  выше среднего  
2789   образовательный кредит        высокий  
19964                 ипотека        высокий  
18598        свадебный кредит        средний  
5241   образовательный кредит        средний  
15611                 ипотека        средний  
11569                 ипотека  выше среднего  
12528              автокредит  выше среднего

In [42]:
df[df['days_employed'] > 0].sample(10)

children  days_employed  dob_years  education  education_id  \
12103         0           1763         34    среднее             1   
9726          0          12141         51    среднее             1   
6754          2           1405         45    среднее             1   
1195          2           1910         39    среднее             1   
17069         0         349633         61    среднее             1   
2854          0           4022         43    среднее             1   
3970          1           2819         31  начальное             3   
16621         0           6275         42     высшее             0   
18042         0           2395         44    среднее             1   
20141         1           7634         51    среднее             1   

          family_status  family_status_id gender  income_type  debt  \
12103  гражданский брак                 1      F    сотрудник     0   
9726          в разводе                 3      F  госслужащий     0   
6754    женат / замужем                 0      M    сотрудник     0   
1195    женат / замужем                 0      F    сотрудник     1   
17069  гражданский брак                 1      F    пенсионер     0   
2854   гражданский брак                 1      M    сотрудник     0   
3970    женат / замужем                 0      M    сотрудник     0   
16621  гражданский брак                 1      M    компаньон     0   
18042   женат / замужем                 0      M    сотрудник     1   
20141  гражданский брак                 1      M  госслужащий     0   

       total_income                                 purpose  \
12103        121699                         сыграть свадьбу   
9726          94640                 приобретение автомобиля   
6754         170589                         свой автомобиль   
1195         115312   операции с коммерческой недвижимостью   
17069        109958                         сыграть свадьбу   
2854          75944  строительство собственной недвижимости   
3970         106196                операции с недвижимостью   
16621        388243                                   жилье   
18042        109862                 покупка жилья для сдачи   
20141        100371                         сыграть свадьбу   

                                                  lemmas  \
12103                          [сыграть,  , свадьба, \n]   
9726                   [приобретение,  , автомобиль, \n]   
6754                           [свой,  , автомобиль, \n]   
1195   [операция,  , с,  , коммерческий,  , недвижимо...   
17069                          [сыграть,  , свадьба, \n]   
2854   [строительство,  , собственный,  , недвижимост...   
3970               [операция,  , с,  , недвижимость, \n]   
16621                                        [жилье, \n]   
18042          [покупка,  , жилье,  , для,  , сдача, \n]   
20141                          [сыграть,  , свадьба, \n]   

               purpose_category   income_level  
12103          свадебный кредит  выше среднего  
9726                 автокредит        средний  
6754                 автокредит  выше среднего  
1195   комерческая недвижимость  выше среднего  
17069          свадебный кредит  выше среднего  
2854                    ипотека        средний  
3970                    ипотека  выше среднего  
16621                   ипотека        высокий  
18042                   ипотека  выше среднего  
20141          свадебный кредит  выше среднего

In [13]:
df[df['days_employed'] > 0]['income_type'].value_counts()

пенсионер      3443
безработный       2
Name: income_type, dtype: int64

In [14]:
result = pd.concat([df[df['days_employed'] > 0],df[df['dob_years']*365 < df['days_employed']]]).drop_duplicates(keep=False)
print(result)

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []


Сравнив две таблицы мы подтверждаем наши предположения: положительное значение опыта полностью соотносится с ошибкой, что опыт превышает кол-во полных лет. Встречается это тех анкетах где тип занятости "пенсионер" или "безработный". Возможная причина данной ошибки в форме заполнения, тк появляется только в том случае, если заемщик не трудоустроен.

Заполним пропуски в столбце days_employed нулями, тк в исследовании они нам не нужны. Наша задача избавиться от пропусков.

In [15]:
df['days_employed'] = df['days_employed'].fillna(0)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Тк в нашем исследовании важны данные по ежемесечному доходу заемщика, то заполним пропуски в столбце total_income медианными значениями по соответствующему роду деятельности.

In [16]:
#data['столбец_где_меняем'] = data.groupby('столбец_группировки')['столбец_где_меняем'].apply(lambda x: x.fillna(x.median()))
df['total_income'] = df.groupby('income_type')['total_income'].apply(lambda x: x.fillna(x.median()))
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Все пропуски заполнены.

Отрицательные значения в столбце children берем по модулю. У нас все еще остаются анкеты с кол-вом детей 20, которые вызывают вопросы, но их мы оставляем без изменения, ткдля обработки этих значений требуется больше входной информации. Со столбцом days_employed поступаем аналогично.

In [17]:
df['children'] = df['children'].abs()
df['days_employed'] = df['days_employed'].abs()

**Вывод**


Те данные которые задействованы в нашем исследовании мы привели в соответствие, стараясь избежать статистических искажений. Некорректные данные которые в нашем исследвании не используются, мы привели к значениям которые позволят нам совершать с датафреймом необходимые действия. 

### Замена типа данных

Переводим значение столбца days_employed от типа float к типу int методом astype('int')

In [18]:
df['days_employed'] = df['days_employed'].astype('int')
df['total_income'] = df['total_income'].astype('int')

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int64 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int64 
 11  purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


### Обработка дубликатов

Как мы уже установили в столбце education мы имеем данные с разным регистром

In [20]:
df['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

ПРиводим данные в столбце к нижнему регистру методом str.lower()

In [21]:
df['education'] = df['education'].str.lower()
df['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Проверим на наличие дубликатов

In [22]:
df.duplicated().sum()

71

Тк в нашем датафрейме отсутствуют уникальные индетификаторы, мы предполагаем что полностью совпадающие строки являются дубликами. Исходя из этого предположения и кол-ва дубликатов, принимаем решение их удалить.

In [23]:
df = df.drop_duplicates().reset_index(drop=True) 
df.duplicated().sum()

0

Дубликаты удалены из таблицы методом drop_duplicates

### Лемматизация

Посмотрим информацию о целях получения кредита. Тк они оформлены в свободной форме, для дальнейшего анализа и категоризации данных проведем лемматизацию.

In [24]:
df.groupby('purpose').count()

children  days_employed  dob_years  \
purpose                                                                      
автомобили                                   478            478        478   
автомобиль                                   494            494        494   
высшее образование                           452            452        452   
дополнительное образование                   460            460        460   
жилье                                        646            646        646   
заняться высшим образованием                 496            496        496   
заняться образованием                        408            408        408   
на покупку автомобиля                        471            471        471   
на покупку подержанного автомобиля           478            478        478   
на покупку своего автомобиля                 505            505        505   
на проведение свадьбы                        768            768        768   
недвижимость                                 633            633        633   
образование                                  447            447        447   
операции с жильем                            652            652        652   
операции с коммерческой недвижимостью        650            650        650   
операции с недвижимостью                     675            675        675   
операции со своей недвижимостью              627            627        627   
покупка жилой недвижимости                   606            606        606   
покупка жилья                                646            646        646   
покупка жилья для сдачи                      651            651        651   
покупка жилья для семьи                      638            638        638   
покупка коммерческой недвижимости            661            661        661   
покупка недвижимости                         621            621        621   
покупка своего жилья                         620            620        620   
получение высшего образования                426            426        426   
получение дополнительного образования        446            446        446   
получение образования                        442            442        442   
приобретение автомобиля                      461            461        461   
профильное образование                       436            436        436   
ремонт жилью                                 607            607        607   
свадьба                                      791            791        791   
свой автомобиль                              478            478        478   
сделка с автомобилем                         455            455        455   
сделка с подержанным автомобилем             486            486        486   
строительство жилой недвижимости             624            624        624   
строительство недвижимости                   619            619        619   
строительство собственной недвижимости       635            635        635   
сыграть свадьбу                              765            765        765   

                                        education  education_id  \
purpose                                                           
автомобили                                    478           478   
автомобиль                                    494           494   
высшее образование                            452           452   
дополнительное образование                    460           460   
жилье                                         646           646   
заняться высшим образованием                  496           496   
заняться образованием                         408           408   
на покупку автомобиля                         471           471   
на покупку подержанного автомобиля            478           478   
на покупку своего автомобиля                  505           505   
на проведение свадьбы                         768           768   
недвижимость                                  63

In [25]:
from pymystem3 import Mystem
from collections import Counter
m = Mystem()

def lemma(row):
    purpose = row['purpose']
    lemmas = m.lemmatize(purpose)
    return lemmas

df['lemmas'] = df.apply(lemma, axis=1)
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose                                lemmas  
0               покупка жилья               [покупка,  , жилье, \n]  
1     приобретение автомобиля     [приобретение,  , автомобиль, \n]  
2               покупка жилья               [покупка,  , жилье, \n]  
3  дополнительное образование  [дополнительный,  , образование, \n]  
4             сыграть свадьбу             [сыграть,  , свадьба, \n]

**Вывод**

С помощью библиотеки pymystem3 мы провели лемматизацию целей кредита.

### Категоризация данных

Для анализа целей мы соедими все леммы в один список lemmas и с помощью функции Counter() проанализируем частоту повторения слов, на основе которых можно вывести категории займа.

In [26]:
lemmas = []
for row in df['lemmas']:
    lemmas.extend(row)

print(Counter(lemmas))

Counter({' ': 33570, '\n': 21454, 'недвижимость': 6351, 'покупка': 5897, 'жилье': 4460, 'автомобиль': 4306, 'образование': 4013, 'с': 2918, 'операция': 2604, 'свадьба': 2324, 'свой': 2230, 'на': 2222, 'строительство': 1878, 'высокий': 1374, 'получение': 1314, 'коммерческий': 1311, 'для': 1289, 'жилой': 1230, 'сделка': 941, 'дополнительный': 906, 'заниматься': 904, 'проведение': 768, 'сыграть': 765, 'сдача': 651, 'семья': 638, 'собственный': 635, 'со': 627, 'ремонт': 607, 'подержанный': 486, 'подержать': 478, 'приобретение': 461, 'профильный': 436})


На основе общей информации и анализа лемм выводи следующие категории займа:

In [27]:
def category(row):

    if 'образование' in row:
        return 'образовательный кредит'
    
    elif 'свадьба' in row:
        return 'свадебный кредит'
    
    elif 'автомобиль' in row:
        return 'автокредит'
    
    elif 'коммерческий' in row:
        return 'комерческая недвижимость'
    
    elif 'ремонт' in row:
        return 'ремонт'
    
    elif 'жилье' in row or 'жилой' in row:
        return 'ипотека'
    
    elif 'строительство' in row:
        if 'жилой'in row or 'свой' in row or 'собственный' in row:
            return 'ипотека'
        return 'кредит на строительство'
    
    elif 'недвижимость' in row:
        return 'ипотека'
        
df['purpose_category'] = df['lemmas'].apply(category)
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose                                lemmas  \
0               покупка жилья               [покупка,  , жилье, \n]   
1     приобретение автомобиля     [приобретение,  , автомобиль, \n]   
2               покупка жилья               [покупка,  , жилье, \n]   
3  дополнительное образование  [дополнительный,  , образование, \n]   
4             сыграть свадьбу             [сыграть,  , свадьба, \n]   

         purpose_category  
0                 ипотека  
1              автокредит  
2                 ипотека  
3  образовательный кредит  
4        свадебный кредит

Тк нам необходимо проанализировать отсутствие задолжностей по кредиту от уровня дохода, нам эти уровни необходимо обозначить. Принимаем следующие значения:

In [28]:
def income_level(income):
    if income <= 50000:
        return 'маленький'
    if (income > 50000) and (income <= 100000):
        return 'средний'
    if (income > 100000) and (income <= 200000):
        return 'выше среднего'
    if (income > 200000) and (income <= 500000):
        return 'высокий'
    if income > 500000:
        return 'очень высокий'

# Создание столбца с категорией

df['income_level'] = df['total_income'].apply(income_level)
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose                                lemmas  \
0               покупка жилья               [покупка,  , жилье, \n]   
1     приобретение автомобиля     [приобретение,  , автомобиль, \n]   
2               покупка жилья               [покупка,  , жилье, \n]   
3  дополнительное образование  [дополнительный,  , образование, \n]   
4             сыграть свадьбу             [сыграть,  , свадьба, \n]   

         purpose_category   income_level  
0                 ипотека        высокий  
1              автокредит  выше среднего  
2                 ипотека  выше среднего  
3  образовательный кредит        высокий  
4        свадебный кредит  выше среднего

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21454 non-null  int64 
 1   days_employed     21454 non-null  int64 
 2   dob_years         21454 non-null  int64 
 3   education         21454 non-null  object
 4   education_id      21454 non-null  int64 
 5   family_status     21454 non-null  object
 6   family_status_id  21454 non-null  int64 
 7   gender            21454 non-null  object
 8   income_type       21454 non-null  object
 9   debt              21454 non-null  int64 
 10  total_income      21454 non-null  int64 
 11  purpose           21454 non-null  object
 12  lemmas            21454 non-null  object
 13  purpose_category  21454 non-null  object
 14  income_level      21454 non-null  object
dtypes: int64(7), object(8)
memory usage: 2.5+ MB


**Вывод**

Все данные по заемщикам категоризированы по уровню дохода и цели получения кредита

## Ответьте на вопросы

### Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [54]:
# функция для построения сводной таблицы и расчета в ней долей просрочек по каждой категории

def correl(column):
    pivot = df.pivot_table(
        index=column, 
        columns= 'debt', 
        values='total_income', 
        aggfunc='count')
    pivot['ratio'] = (pivot[1] / (pivot[1] + pivot[0])*100).round(2)
    return pivot

pd.options.display.float_format = '{:,.2f}'.format

In [55]:
correl('children')

debt             0        1  ratio
children                          
0        13,028.00 1,063.00   7.54
1         4,410.00   445.00   9.17
2         1,858.00   194.00   9.45
3           303.00    27.00   8.18
4            37.00     4.00   9.76
5             9.00      NaN    NaN
20           68.00     8.00  10.53

In [56]:
correl('children')['ratio'].sort_values(ascending = False)

children
20   10.53
4     9.76
2     9.45
1     9.17
3     8.18
0     7.54
5      NaN
Name: ratio, dtype: float64

**Вывод**

Зависимость есть, заемщики без детей реже имеют задолжности по выплатам кредита. Данные по заемщикам с 5 детьми не репрезентативны, тк выборка слишком мала.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [57]:
correl('family_status')

debt                       0    1  ratio
family_status                           
Не женат / не замужем   2536  274   9.75
в разводе               1110   85   7.11
вдовец / вдова           896   63   6.57
гражданский брак        3763  388   9.35
женат / замужем        11408  931   7.55

In [58]:
correl('family_status')['ratio'].sort_values(ascending = False)

family_status
Не женат / не замужем   9.75
гражданский брак        9.35
женат / замужем         7.55
в разводе               7.11
вдовец / вдова          6.57
Name: ratio, dtype: float64

**Вывод**

Заемщики не состоящие/состоявшие в официальных отношениях чаще других имеют задолжности по выплатам по кредиту.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [59]:
correl('income_level')

debt               0     1  ratio
income_level                     
высокий         4501   344   7.10
выше среднего  10895  1029   8.63
маленький        349    23   6.18
очень высокий    208    14   6.31
средний         3760   331   8.09

In [60]:
correl('income_level')['ratio'].sort_values(ascending = False)

income_level
выше среднего   8.63
средний         8.09
высокий         7.10
очень высокий   6.31
маленький       6.18
Name: ratio, dtype: float64

**Вывод**

Заемщики с низким и очень высоким уровнем дохода реже других берут кредиты, поэтому их выборка невелика. Делать однозначные выводы опираясь на это трудно. Чаще всех кредиты берут люди с доходом выше среднего. Можно сделать предположение о высокой закредитованности этой группы, что может приводить к задолжностям по выплатам.

### Как разные цели кредита влияют на его возврат в срок?

In [61]:
correl('purpose_category')

debt                         0    1  ratio
purpose_category                          
автокредит                3903  403   9.36
ипотека                   7680  594   7.18
комерческая недвижимость  1212   99   7.55
кредит на строительство    565   54   8.72
образовательный кредит    3643  370   9.22
ремонт                     572   35   5.77
свадебный кредит          2138  186   8.00

In [62]:
correl('purpose_category')['ratio'].sort_values(ascending = False)

purpose_category
автокредит                 9.36
образовательный кредит     9.22
кредит на строительство    8.72
свадебный кредит           8.00
комерческая недвижимость   7.55
ипотека                    7.18
ремонт                     5.77
Name: ratio, dtype: float64

In [73]:
for i in df['income_level'].unique():
    display(df.query('income_level == @i').groupby(['income_level','purpose_category'])\
          .agg({'total_income':'count'}).sort_values(by='total_income'))

total_income
income_level purpose_category                      
высокий      кредит на строительство            121
             ремонт                             139
             комерческая недвижимость           293
             свадебный кредит                   526
             образовательный кредит             867
             автокредит                         993
             ипотека                           1906

total_income
income_level  purpose_category                      
выше среднего ремонт                             349
              кредит на строительство            350
              комерческая недвижимость           715
              свадебный кредит                  1289
              образовательный кредит            2263
              автокредит                        2409
              ипотека                           4549

total_income
income_level purpose_category                      
средний      ремонт                             106
             кредит на строительство            133
             комерческая недвижимость           262
             свадебный кредит                   450
             образовательный кредит             768
             автокредит                         790
             ипотека                           1582

total_income
income_level  purpose_category                      
очень высокий ремонт                               6
              кредит на строительство              7
              комерческая недвижимость            18
              свадебный кредит                    21
              образовательный кредит              34
              автокредит                          40
              ипотека                             96

total_income
income_level purpose_category                      
маленький    ремонт                               7
             кредит на строительство              8
             комерческая недвижимость            23
             свадебный кредит                    38
             автокредит                          74
             образовательный кредит              81
             ипотека                            141

**Вывод**

Ипотечное кредитование является самым востребованным из представленных. При этом у него низкий процент наличия задолжностей, что может обьясняться жесткой политикой банков относительно ипотеки. Так же низкий процент задолжностей при займе на ремонт, что может быть обьяснено высокой процентной ставкой по потребительским кредитам и желании заемщика быстрее закрыть его. Автокредит имеет самый высокий процент по задолжностям. Можно предположить что это обьясняется культурой владения автомобиля и роли статусной вещи, что подталкивает людей брать крдеиты на машины которые они не могут себе позволить при разумном потреблении. Образовательный кредит так же имеет высокий процент задолжностей, что может обьясняться стоимостью образвания, и не всегда быстрому результату в виде повышения зп от получения образования/повышения квалификации.

Так же отмечаем, что распределение по целям кредитования почти одно и то же для всех уровней дохода.

## Общий вывод

In [63]:
correl('gender')

debt           0      1  ratio
gender                        
F      13,180.00 994.00   7.01
M       6,532.00 747.00  10.26
XNA         1.00    NaN    NaN

Так же мы видим что женщины имеют меньший процент задолжностей по кредиту.
Посмотрим общую информацию по переработанному датафрейму с отсутствием задолжностей и попробуем выделить портрет идеального заемщика.

In [53]:
df_good = df[df['debt'] == 0]
for col in df_good.columns:
    print(str(col))
    display(df_good[col].value_counts())

children


0     13028
1      4410
2      1858
3       303
20       68
4        37
5         9
Name: children, dtype: int64

days_employed


0         1933
133         16
327         15
204         14
223         14
          ... 
6682         1
344186       1
375310       1
371212       1
6874         1
Name: days_employed, Length: 8785, dtype: int64

dob_years


35    569
40    559
41    556
42    554
34    537
38    533
39    528
33    525
36    511
44    504
48    502
37    484
31    483
30    483
29    479
49    477
43    468
50    466
32    457
45    456
52    456
56    450
28    446
54    445
27    444
47    440
46    434
58    430
57    426
53    424
51    422
55    420
59    413
26    361
60    352
61    344
62    330
25    314
63    257
64    248
24    240
23    232
65    179
66    174
67    158
22    158
21     97
0      93
68     91
69     81
70     62
71     55
20     47
72     31
19     13
73      8
74      6
75      1
Name: dob_years, dtype: int64

education


среднее                13808
высшее                  4972
неоконченное высшее      676
начальное                251
ученая степень             6
Name: education, dtype: int64

education_id


1    13808
0     4972
2      676
3      251
4        6
Name: education_id, dtype: int64

family_status


женат / замужем          11408
гражданский брак          3763
Не женат / не замужем     2536
в разводе                 1110
вдовец / вдова             896
Name: family_status, dtype: int64

family_status_id


0    11408
1     3763
4     2536
3     1110
2      896
Name: family_status_id, dtype: int64

gender


F      13180
M       6532
XNA        1
Name: gender, dtype: int64

income_type


сотрудник          10023
компаньон           4702
пенсионер           3613
госслужащий         1371
предприниматель        2
студент                1
безработный            1
Name: income_type, dtype: int64

debt


0    19713
Name: debt, dtype: int64

total_income


142594    972
172357    472
118514    352
150447    138
97012       3
         ... 
97178       1
215964      1
185247      1
142240      1
126820      1
Name: total_income, Length: 17154, dtype: int64

purpose


свадьба                                   727
сыграть свадьбу                           707
на проведение свадьбы                     704
операции с недвижимостью                  620
покупка коммерческой недвижимости         614
операции с жильем                         604
жилье                                     600
покупка жилья для сдачи                   599
операции с коммерческой недвижимостью     598
покупка жилья                             598
строительство собственной недвижимости    593
покупка жилья для семьи                   593
недвижимость                              591
покупка своего жилья                      586
покупка недвижимости                      578
операции со своей недвижимостью           577
строительство жилой недвижимости          576
ремонт жилью                              572
покупка жилой недвижимости                565
строительство недвижимости                565
на покупку своего автомобиля              459
заняться высшим образованием      

lemmas


TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 4588, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[автомобиль, \n]                                          886
[свадьба, \n]                                             727
[сыграть,  , свадьба, \n]                                 707
[на,  , проведение,  , свадьба, \n]                       704
[операция,  , с,  , недвижимость, \n]                     620
[покупка,  , коммерческий,  , недвижимость, \n]           614
[операция,  , с,  , жилье, \n]                            604
[жилье, \n]                                               600
[покупка,  , жилье,  , для,  , сдача, \n]                 599
[операция,  , с,  , коммерческий,  , недвижимость, \n]    598
[покупка,  , жилье, \n]                                   598
[покупка,  , жилье,  , для,  , семья, \n]                 593
[строительство,  , собственный,  , недвижимость, \n]      593
[недвижимость, \n]                                        591
[покупка,  , свой,  , жилье, \n]                          586
[покупка,  , недвижимость, \n]                            578
[операци

purpose_category


ипотека                     7680
автокредит                  3903
образовательный кредит      3643
свадебный кредит            2138
комерческая недвижимость    1212
ремонт                       572
кредит на строительство      565
Name: purpose_category, dtype: int64

income_level


выше среднего    10895
высокий           4501
средний           3760
маленький          349
очень высокий      208
Name: income_level, dtype: int64

- без детей
- замужем
- пол: женщина
- цель: ипотека
- доход выше среднего

Что полностью соотноситься с полученными выводами